### ISSUE TITLE
UI efficiency improvement - edit log image

### ISSUE THREAD

**Issue by andrixnet** [cgeo/cgeo#14514]
### Tell us your idea!

Currently the UI for adding a log image is reasonable, with the necessary fields and works ok. 

Logging multiple finds and adding images to each of them I discovered the UI makes it somewhat cumbersome. 
My use case was simple: load an already existing image (photos taken earlier), add a caption, end.
The UI requires me to either scroll or close the keyboard to finalize this form, because the "OK" button is hidden by the keyboard. 

Keyboard open, covering "OK" button.
![Screenshot_20230814-160705_cgeo](https://github.com/cgeo/cgeo/assets/3382777/339ba99d-b473-43ab-9fff-dbea7cff5b39)

Keyboard closed, "OK" button visible.
![Screenshot_20230814-160711_cgeo](https://github.com/cgeo/cgeo/assets/3382777/0e777004-a774-4b5e-aca6-13dd136e2c95)

### What solution would you suggest?

Since the most often use case is only edit the caption, it would be more efficient to have access to the "OK" button above the area of the screen where the soft keyboard appears. 

My suggestion would be above the "existing" and "new" buttons, with the logic that most action buttons in c:geo are on the top part of the screen.

### Additional Information

User desiring to edit description still has scroll or "next" button on keyboard.
User desiring to change image size still has scroll. 
But these controls I believe are much less often accessed.

**Comment by ztNFny** [comment]
The 3 buttons could be moved to a toolbar, this would match the UI we have in log screen

**Comment by fm-sys** [comment]
> The 3 buttons could be moved to a toolbar, this would match the UI we have in log screen

...waypoints, personal note and many other places. Sounds consistent - like it!

**Comment by eddiemuc** [comment]
I agree with the request. I believe that the "edit (log) image" dialog could use a complete UX makeover.

Considering the new image edit functions requested in #14515, the "edit image" dialog needs to provide the following UX elements:

* Central Element: the image itself
* Additional edit attributes:
  * caption
  * description
  * Resolution selection
* Actions on image
   * select another image from camera (today a text button with label "new")
   * select  another image from storage (today a text button with label "existing")  
   * rotate left / right / 180 degree
   * flip horizontal / vertical
   * crop (thats optional, but I would find it nice...)
* Actions closing the dialog:
  * OK: apply changes to (log) image
  * Cancel: revoke all changes and close dialog
  * Delete: delete the image (in log image list) and close dialog

That's quite a lot of stuff to put in a dialog. Would be nice to have a complete target UX idea for this dialog before starting to code something.

**Comment by eddiemuc** [comment]
(edited my last comment)

**Comment by ztNFny** [comment]
I'm against implementing all that image editing functionality. That's something the gallery app already does quite well.
That said a way to easily pass the image to gallery and get the edited result would be welcome.

**Comment by eddiemuc** [comment]
Ok, here is my UX redesign suggestion for discussion:
* make the dialog a pure "image edit" dialog. Images can only be added/removed via the "Log image list" in the log dialog. This means:
   * remove "delete" button 
   * removing the "attach image" function in the log's dialog context menu
* remove "cancel" and "ok" buttons. "cancel" can be done via "arrow-left" in upper-left and "ok" via "hook" in the upper right
* divide edit functions in three categories "basic", "description" and "image". 
    * "Basic lets you edit "caption and "scaling"
    * "Description" edits description
    * "Image" provides image editing options (select new from storage/camera, rotate, flip, edit extermal)

Could look e.g. like the following screenshots:

![image](https://github.com/cgeo/cgeo/assets/6909759/019f1b23-43c5-4fee-b70e-ae5d235a01ea)
![image](https://github.com/cgeo/cgeo/assets/6909759/9ca6fec2-8f6f-47ad-a830-7828cc5b99f5)
![image](https://github.com/cgeo/cgeo/assets/6909759/071c1af9-3257-4542-b6e3-6808549db07b)

**Comment by eddiemuc** [comment]
As time permits and if no one objects I would tackle this next, going with the design proposed before.

**Comment by fm-sys** [comment]
Hm, I wouldn't use the ribbon design. While this is common in MS office, I never have seen something like that on Android before. _If_ we go that way, use material tabs instead. Or just put all options below each other. Maybe with line separators similar to our cache listing activity. The second is IMHO the most intuitive one.

**Comment by eddiemuc** [comment]
> Or just put all options below each other. Maybe with line separators similar to our cache listing activity. 

Wouldn't that be like it is today? My idea was to go for a one-screen-view where no scrolling is necessary any more

**Comment by ztNFny** [comment]
With todays phone sizes I think even if put below each other scrolling wouldn't be required.
I'd suggest to order them:
- image preview
- editing tools
- scaling dropdown
- title
- description
Like that even on small phones title should be visible by default and description is seldomly used anyway

But I can live with both ways

**Comment by eddiemuc** [comment]
Fine for me. Also, this makes the change much easier :-)

**Comment by fm-sys** [comment]
"Now I have to click one more time instead of scroll" doesn't sounds like a big improvement 😆

IMHO the current look isn't that bad. Moving ok/cancel to the action bar, show just "edit image" instead of the cutoff cache title in the action bar, removing the "add images" buttons and some rearrangement and the dialog is just perfect. 

BTW, on my device (6" which is quite the average these days) I don't have to scroll at all so the "improvement" would make things worse for me. 

[KISS](https://de.m.wikipedia.org/wiki/KISS-Prinzip) - "Keep it simple, stupid!"

![Screenshot_20230828_085639_cgeo geocaching_copy_270x585](https://github.com/cgeo/cgeo/assets/64581222/6bdc5cdb-326d-4d6f-ba10-e100d386e576)

But I've noticed another way more important thing: Rotating the image currently takes more than a second on my phone - this is totally unacceptable! Rotating a `View` component is an easy and performant operation, while applying rotation to the image itself seems to be not. So we should just store the user selected rotation and apply it to the image while upload.

**Comment by eddiemuc** [comment]
PR #14561 will change the image edit gui to the layout discussed here

> But I've noticed another way more important thing: Rotating the image currently takes more than a second on my phone - this is totally unacceptable! Rotating a View component is an easy and performant operation, while applying rotation to the image itself seems to be not. So we should just store the user selected rotation and apply it to the image while upload.

@fm-sys : this would be a valid concern, although I can't reproduce it on my AVD. In any case, since there's no approval yet whether to leave the rotate/flip (or the "edit external") in c:geo I would rather not address this issue right now - in  case the feature gets removed again due to team decision. I haven't forgotten your request for animated rotate/flip as well :-) But I would like to separate it from the GUI changes in #14561 (while at the same point having to wait for a merge of this PR due to code structure changes which would otherwise lead to merge conflicts). Hope that's ok for you. Should we open a separate issue for the rotate/flip thingy or maybe move discussion to  #14515

**Comment by eddiemuc** [comment]
@fm-sys : sorry I couldn't resist to add animation after all to the PR... Your performance problem should also be fixed with this. If rotate/flip is to be removed from c:geo that's still fine for me.

**Comment by fm-sys** [comment]
> @fm-sys : sorry I couldn't resist to add animation after all to the PR... Your performance problem should also be fixed with this. If rotate/flip is to be removed from c:geo that's still fine for me.

No need to say sorry, really appreciated 😃👍

**Comment by fm-sys** [comment]
Here is a video which shows the performance problems while rotation. Visually it got significantly better with the animation, however it's still kinda slow. 

https://github.com/cgeo/cgeo/assets/64581222/efeb09c6-90fe-47be-be4b-b3f6e34bd6f9

This is the image for testing purposes:

![IMG_20230828_061652](https://github.com/cgeo/cgeo/assets/64581222/ca43f335-49b0-406f-84d3-f1eb5ae33a01)

**Comment by eddiemuc** [comment]
I don't have my IDE available (vacation). However, it seems clear time that the lag comes from the saving of the modified image to disk - which seems to take a long time on your device or with this image. Is this a very large image?

The process could further be improved by not saving the modified image at all or only when absolutely necessary. This is when the dialog is closed or when the image shall be passed to external editor. 

Or by installing a background thread constantly watching for image changes and saving them one after another

**Comment by fm-sys** [comment]
Well, it's 7.2 MB but this shouldn't take _that_ long nowadays... I was more thinking about a performance issues in the rotation algorithm, but who knows. Should be investigated further...

**Comment by ztNFny** [comment]
wouldn't it be enough to set the orientation flag in images metadata instead of actually rotating them?

**Comment by fm-sys** [comment]
Addition: Flipping the image takes more or less the same amount of time.

Bug: I just noticed, that after using the flip feature, the rotation button turns the image in the opposite direction. I doubt this is intended ;-)

Finally: Just want to say that I really love the new design of the image edit dialog. It looks so clean and tidy! ❤️
If only all of c:geo could look like this 🤩

**Comment by eddiemuc** [comment]
> wouldn't it be enough to set the orientation flag in images metadata instead of actually rotating them?

Probably, if the issue is with the rotation Algo. But how can this be done? Does our metadata lib support this?

**Comment by eddiemuc** [comment]
> > wouldn't it be enough to set the orientation flag in images metadata instead of actually rotating them?
> 
> Probably, if the issue is with the rotation Algo. But how can this be done? Does our metadata lib support this?

Update: I looked into ExifInterface (the android class we use to read e.g. orientation). It support setAttribute saveAttributes methods, but the save method is supposed to be expensive because it creates a copy of the underlying file as well. Also I wonder how the class can write to the image file if it was opened using an inputstream... It looks like this has to be tried out

**Comment by eddiemuc** [comment]
> Bug: I just noticed, that after using the flip feature, the rotation button turns the image in the opposite direction. I doubt this is intended ;-)

That's strange. I had this bug during development and thought I fixed it...

**Comment by eddiemuc** [comment]
@fm-sys : a question: when you open the edit image dialogand then immediately tap "edit external" action (without rotating or flipping first), do you also have a noticeable delay there? (If yes this would hint towards a problem with long saving times because the dialog will make an internal new copy of the image to pass for edit)

**Comment by fm-sys** [comment]
> @fm-sys : a question: when you open the edit image dialogand then immediately tap "edit external" action (without rotating or flipping first), do you also have a noticeable delay there?

Not really a noticeable delay. But I'm not absolutely sure...

> In any case, I can try this out earliest in a week unfortunately. If this bums you out too much, please deactivate rotation and flipping by setting the buttons visibility to gone. Sorry I can't check this quicker.

Totally fine. It's still way better than nothing, and as long as we don't start a new beta nightly users are used to unfinished features :)

**Comment by eddiemuc** [comment]
I was able to take a first look, some updates:
- to my surprise, the lagging part is indeed the rotation/flip algorithm using Android Matrix (not File operations)
- trying to follow @ztNFny s idea to just set the exit orientation byte instead of actually rotating/flipping. This has additional advantage that image file doesn't need to be loaded into memory to perform the operation. I encounter some problems using ExifInterface saveAttributes but hope to overcome those. 
- Working on this I also discovered that c:geo itself does not handle all orientation exit flags correctly. I will fix this in the process
- even when doing the "exif" way, rewriting the file is still a heavy operation. I will thua restrict this to where absolutely necessary


One point remains: the root problem is the size of the image file (6-7MB) and it's pixel size. It seems wasteful to perform operations on such a big file only to scale it down later when uploading it. This is not only about compute time but also about disk space. IIRC then GC.com has a size limit of about 2MB? Would it make sense to pre-scale all images down to e.g. 2048px max already on selecting them?
